File: es7/es7.py
#!/usr/bin/env python3

import sys
import string

class Task(object):
    def __init__(self, name):
        self.name = name
        self.assigned = False
        self.completed = False
        self.depends = []

    def pos(self):
        return string.ascii_uppercase.index(self.name) - string.ascii_uppercase.index('A')

class TaskPool(object):
    def __init__(self):
        self.tasks = {}

    def add_rule(self, taskname, dependson):
        task = self.tasks.get(taskname)
        if not task:
            task = Task(taskname)
            self.tasks[taskname] = task

        dep = self.tasks.get(dependson)
        if not dep:
            dep = Task(dependson)
            self.tasks[dependson] = dep

        task.depends.append(dep)

    def find_available(self):
        avail = None
        for task in self.tasks.values():
            if task.assigned or task.completed:
                continue

            found = True
            for d in task.depends:
                if not d.completed:
                    found = False
                    break

            if found and (avail is None or task.name < avail.name):
                avail = task

        return avail


class Worker(object):
    def __init__(self):
        self.assigned_task = None
        self.end = 0

class WorkPool(object):
    def __init__(self, size):
        self.workers = [Worker() for i in range(size)]
        self.busy_workers = 0
        self.tasks_done = []

    def get_worker(self):
        for w in self.workers:
            if w.assigned_task is None:
                return w
        return None

    def assign_task(self, task, end):
        w = self.get_worker()
        if w:
            task.assigned = True
            w.assigned_task = task
            w.end = end
            self.busy_workers += 1
            return True

        return False

    def find_next_end(self):
        end = 10000000000
        for w in self.workers:
            if w.assigned_task and w.end < end:
                end = w.end
        return end

    def process(self, now):
        for w in self.workers:
            if w.assigned_task and w.end == now:
                self.tasks_done.append(w.assigned_task)
                self.busy_workers -= 1
                w.assigned_task.completed = True
                w.assigned_task = None

    def print(self, now):
        values = []
        values.append("{:4} ".format(now))

        for w in self.workers:
            if w.assigned_task:
                values.append(" {} ".format(w.assigned_task.name))
            else:
                values.append(" . ")

        for t in self.tasks_done:
            values.append("{}".format(t.name))

        print("".join(values))

if __name__ == "__main__":
    if len(sys.argv) < 3:
        print("Usage: {} <filename> <numworkers>".format(sys.argv[0]), file = sys.stderr)
        sys.exit(-1)

    try:
        wsize = int(sys.argv[2])
    except:
        print("Cannot parse the number of workers: {}".format(sys.argv[2]), file = sys.stderr)
        sys.exit(-1)

    wp = WorkPool(wsize)
    tp = TaskPool()
    try:
        with open(sys.argv[1], "r") as f:
            for line in f:
                a = line.split(" ")
                required, task = a[1], a[7]
                tp.add_rule(task, required)
    except:
        print("Cannot open {} for reading".format(sys.argv[1]), file = sys.stderr)
        sys.exit(-1)

    now = 0
    while True:
        while True:
            task = tp.find_available()
            if task is None:
                break

            end = now + task.pos() + 61
            if not wp.assign_task(task, end):
                break

        wp.print(now)

        if wp.busy_workers == 0:
            break

        now = wp.find_next_end()
        wp.process(now)

    print("Seconds: {}".format(now))

File: es7/es7.go
package main

import (
	"fmt"
	"os"
	"strconv"
)

const MaxInt = int(^uint(0) >> 1)

type Task struct {
	Name      int
	Assigned  bool
	Completed bool
	Depends   []*Task
}

type TaskPool struct {
	Tasks map[int]*Task
}

func (tp *TaskPool) AddRule(taskName, dependsOn int) {
	task, ok := tp.Tasks[taskName]
	if !ok {
		task = &Task{Name: taskName}
		tp.Tasks[taskName] = task
	}

	dep, ok := tp.Tasks[dependsOn]
	if !ok {
		dep = &Task{Name: dependsOn}
		tp.Tasks[dependsOn] = dep
	}
	task.Depends = append(task.Depends, dep)
}

func (tp *TaskPool) FindAvailable() *Task {
	var avail *Task
	for _, task := range tp.Tasks {
		if task.Assigned || task.Completed {
			continue
		}

		found := true
		for _, dep := range task.Depends {
			if !dep.Completed {
				found = false
			}
		}

		if found && (avail == nil || task.Name < avail.Name) {
			avail = task
		}
	}
	return avail
}

func makeTaskPool() TaskPool {
	return TaskPool{Tasks: map[int]*Task{}}
}

type Worker struct {
	Name         int
	AssignedTask *Task
	End          int
}

type WorkPool struct {
	Workers     []*Worker
	BusyWorkers int
	TasksDone   []*Task
}

func (wp *WorkPool) GetWorker() *Worker {
	for _, w := range wp.Workers {
		if w.AssignedTask == nil {
			return w
		}
	}
	return nil
}

func (wp *WorkPool) AssignTask(task *Task, end int) bool {
	w := wp.GetWorker()
	if w != nil {
		wp.BusyWorkers++
		w.AssignedTask = task
		w.End = end
		task.Assigned = true
		return true
	}
	return false
}

func (wp *WorkPool) FindMinEnd() int {
	end := MaxInt
	for _, w := range wp.Workers {
		if w.AssignedTask != nil && w.End < end {
			end = w.End
		}
	}
	return end
}

func (wp *WorkPool) Process(now int) {
	for _, w := range wp.Workers {
		if w.AssignedTask != nil && w.End == now {
			wp.TasksDone = append(wp.TasksDone, w.AssignedTask)
			wp.BusyWorkers--
			w.AssignedTask.Completed = true
			w.AssignedTask = nil
		}
	}
}

func (wp *WorkPool) Print(now int) {
	fmt.Printf("%4d", now)
	for _, w := range wp.Workers {
		if w.AssignedTask != nil {
			fmt.Printf(" %c ", w.AssignedTask.Name)
		} else {
			fmt.Printf(" . ")
		}
	}
	for _, t := range wp.TasksDone {
		fmt.Printf("%c", t.Name)
	}
	fmt.Println()
}

func makeWorkPool(size int) WorkPool {
	workers := []*Worker{}
	for i := 0; i < size; i++ {
		workers = append(workers, &Worker{
			Name:         i,
			AssignedTask: nil,
			End:          0,
		})
	}
	return WorkPool{
		Workers:   workers,
		TasksDone: []*Task{},
	}
}

func main() {
	val := mainWithInt()
	if val != 0 {
		os.Exit(val)
	}
}

func mainWithInt() int {
	if len(os.Args) < 3 {
		fmt.Fprintf(os.Stderr, "Usage %s [filename] [numworkers]\n", os.Args[0])
		return -1
	}

	wsize, err := strconv.ParseInt(os.Args[2], 10, 64)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Cannot parse the number of workers\n")
		return -1
	}

	input, err := os.Open(os.Args[1])
	if err != nil {
		fmt.Fprintf(os.Stderr, "Cannot open %s for reading\n", os.Args[1])
		return -1
	}
	defer input.Close()

	tp := makeTaskPool()
	wp := makeWorkPool(int(wsize))

	var requisite, task int
	for {
		_, err := fmt.Fscanf(input, "Step %c must be finished before step %c can begin.\n",
			&requisite, &task)
		if err != nil {
			break
		}

		tp.AddRule(task, requisite)
	}

	now := 0
	for {
		for {
			task := tp.FindAvailable()
			if task == nil {
				break
			}
			end := now + task.Name - 'A' + 61
			if !wp.AssignTask(task, end) {
				break
			}
		}

		wp.Print(now)

		if wp.BusyWorkers == 0 {
			break
		}

		now = wp.FindMinEnd()
		wp.Process(now)
	}

	fmt.Println("Seconds:", now)

	return 0
}
